### PR TITLE
z/OS port provisioning and resource monitoring for ports

### PR DESCRIPTION
**Pull request by lewisja1406**
This is specifically to z/OS at the moment. It will need abstracting for generic ports.

### CODE REVIEW

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosManagerImpl.java [hunk 1]
```diff
@@ -221,6 +222,16 @@ public void provisionGenerate() throws ManagerException, ResourceUnavailableExce
             dependencyTags.remove("PRIMARY");
         }
 
+        System.out.println("LEWO");
+        System.out.println(cps.getAllProperties());
+        try {
+			System.out.println(cps.getProperty("image", "ports", "MV2D"));
+		} catch (ConfigurationPropertyStoreException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Comment (hobbit1983) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosManagerImpl.java:231]: need to re-throw an exception

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosManagerImpl.java [hunk 2]
```diff
@@ -680,4 +696,8 @@ public String getZosConnectInstallDir(ZosBaseImageImpl image) throws ZosManagerE
         }
         return this.zosConnectInstallDir;
```
Comment (hobbit1983) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosManagerImpl.java:697]: zosConnect????

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPoolPorts.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	public ZosPoolPorts(ZosManagerImpl manager, IDynamicStatusStoreService dss, IResourcePoolingService rps) throws ZosManagerException {
+		this.manager = manager;
+		this.dss = dss;
+		this.rps = rps;
+	}
+	
+	public String allocatePort(String image) throws ZosManagerException {
+
+		// The the pool of ports for the specific image
+		List<String> resourceStrings = PoolPorts.get(image);
```
Comment (hobbit1983) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPoolPorts.java:35]: what happens if there are no resource strings for this image?

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPoolPorts.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		
+		String thePort = null;
+		try {
+			List<String> zosPorts = this.rps.obtainResources(resourceStrings, null, 1, 1, dss, "zosport");
```
Comment (hobbit1983) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPoolPorts.java:39]: if this is empty after the call to obtain resources then thePort will be returned as null

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosResourceManagement.java [hunk 5]
```diff
@@ -38,19 +39,20 @@ public boolean initialise(IFramework framework, IResourceManagement resourceMana
...
-                20, 
-                TimeUnit.SECONDS);
-    }
+	@Override
+	public void start() {
+		this.resourceManagement.getScheduledExecutorService().scheduleWithFixedDelay(slotResourceMonitor,
+				this.framework.getRandom().nextInt(20), 20, TimeUnit.SECONDS);
+		
+		this.resourceManagement.getScheduledExecutorService().scheduleWithFixedDelay(zosPortResourceMonitor,
+				this.framework.getRandom().nextInt(20), 20, TimeUnit.SECONDS);
```
Comment (hobbit1983) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosResourceManagement.java:54]: point for discussion - should the timeout be configurable through the CPS?

**Comment by lewisja1406** [comment]
Review changes made please re-review

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+		
+		try {
+			
+			// Get the list of ports allocated in the DSS
+			Map<String, String> zosPortsInDss = dss.getPrefix("zosport");
+			
+			Set<String> allActiveRuns = this.framework.getFrameworkRuns().getActiveRunNames();
+			
+			// Iterate through all z/OS ports stored in the DSS
+			for (String prefix : zosPortsInDss.keySet()) {
```
Comment (Jimbo4794) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java:45]: its a key not prefix

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	
+				// Delete DSS port allocation to run not active
+				if (!allActiveRuns.contains(zosPortsInDss.get(prefix))) {
+					String removeRun = zosPortsInDss.get(prefix);
+					String removePort = prefix.split("\\.")[1];
+					
+					logger.info("Found inactive run " + removeRun + ". Freeing port " + removePort);
```
Comment (Jimbo4794) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java:52]: Dont message about the run, not your job, message youve found a dirty port on image X

File: galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+			}
+		} catch (Exception e) {
+			logger.error("Failure during scanning DSS for z/OS ports");
+		}		
+		
+        this.resourceManagement.resourceManagementRunSuccessful();
+        logger.info("Finished cleaning up z/OS port allocation");
+	}
+	
+	public void runFinishedOrDeleted(String runName) {
```
Comment (Jimbo4794) [inline galasa-managers-parent/galasa-managers-zos-parent/dev.galasa.zos.manager/src/main/java/dev/galasa/zos/internal/ZosPortResourceMonitor.java:65]: Any similiar above change for this too

**Review by Jimbo4794 — CHANGES_REQUESTED** [review]

**Review by hobbit1983 — APPROVED** [review]

**Comment by lewisja1406** [comment]
@Jimbo4794 Please can you re-review based on the changes. I have tested and things look good

**Review by Jimbo4794 — APPROVED** [review]

**Pull request merged**